### PR TITLE
adding a new flag --mode for dev and pro 

### LINKED ISSUE #6575: Custom Env Variables are invalid in version 2.1
### What version of `astro` are you using?

2.1.0

### Are you using an SSR adapter? If so, which one?

node standalone

### What package manager are you using?

pnpm

### What operating system are you using?

Windows

### What browser are you using?

Chrome

### Describe the Bug

Custom .env.test are invalid in vCustom environment variables are invalid in version 2.1, such as .env.test
execute `cross-env NODE_ENV=test astro build --mode test`. But the variables in the .env.production file are still used

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-ueaw7y-s1mlqj?file=package.json

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by wulinsheng123**
## Changes
Fix #6575
adding a new --mode flag for dev and pro.
such as dev `pnpm run dev --mode=test`
such as build `pnpm run pnpm --mode=test`

### CODE REVIEW

File: packages/astro/test/astro-mode-envs.test.js
Comment (ematipico) [inline packages/astro/test/astro-mode-envs.test.js:15]:
```suggestion
	describe('when running "astro build"', () => {
```

File: packages/astro/test/astro-mode-envs.test.js
Comment (ematipico) [inline packages/astro/test/astro-mode-envs.test.js:20]:
```suggestion
		it('should render public env and private env', async () => {
```

File: packages/astro/test/astro-mode-envs.test.js
Comment (ematipico) [inline packages/astro/test/astro-mode-envs.test.js:29]:
```suggestion
	describe('when running "astro dev"', () => {
```

File: packages/astro/test/astro-mode-envs.test.js
Comment (ematipico) [inline packages/astro/test/astro-mode-envs.test.js:39]:
```suggestion
		it('should render public env and private env', async () => {
```

**Review by ematipico — CHANGES_REQUESTED** [review]
Things that we should cover before merging the PR:
- add the new flag to the CLI (help section);
- make some manual testing check via CLI to make sure it works;
- make sure that this new flag is documented in `docs.astro.build`

**Comment by wulinsheng123** [comment]
> Things that we should cover before merging the PR:
> 
> * make some manual testing check via CLI to make sure it works;

What are you referring to recording a video?

File: packages/astro/src/core/dev/container.ts
Comment (wulinsheng123) [reply packages/astro/src/core/dev/container.ts:161]: image
I did what you said. but it will give me a ts error.

File: .changeset/odd-foxes-invent.md
Comment (ematipico) [reply .changeset/odd-foxes-invent.md]: As a reference, you can read this bullet point: https://semver.org/#spec-item-7

In your PR, we are adding a **new functionality** that is backwards compatible, meaning we are not breaking existing code that users use. The API that is mentioned on the website, it's our CLI in this case.

**Comment by ematipico** [comment]
> > Things that we should cover before merging the PR:
> > 
> > * make some manual testing check via CLI to make sure it works;
> 
> What are you referring to recording a video?

It's not necessary unless you want to! Just a mention that you manually tested your new change I think it's enough! 

We don't test the CLI directly in our testing suite, so it's always best to test it locally to ensure it works as expected.

**Comment by wulinsheng123** [comment]
> 



> > > Things that we should cover before merging the PR:
> > > 
> > > * make some manual testing check via CLI to make sure it works;
> > 
> > 
> > What are you referring to recording a video?
> 
> It's not necessary unless you want to! Just a mention that you manually tested your new change I think it's enough!
> 
> We don't test the CLI directly in our testing suite, so it's always best to test it locally to ensure it works as expected.

I got it, of course I did

File: .changeset/odd-foxes-invent.md
Comment (wulinsheng123) [reply .changeset/odd-foxes-invent.md]: > As a reference, you can read this bullet point: https://semver.org/#spec-item-7
> 
> In your PR, we are adding a **new functionality** that is backwards compatible, meaning we are not breaking existing code that users use. The API that is mentioned on the website, it's our CLI in this case.

Thanks for your answer.
I will  read it late

**Comment by ematipico** [comment]
> I got it, of course I did

Good to know :) next time you can add this information in the template, under the "Testing" section

File: .changeset/odd-foxes-invent.md
Comment (matthewp) [inline .changeset/odd-foxes-invent.md:2]: this should be minor.

**Review by matthewp — CHANGES_REQUESTED** [review]
This would be a minor change.

File: package.json
Comment (wulinsheng123) [reply package.json:44]: no problem.

**Comment by wulinsheng123** [comment]
> Looks good to me! We keep it blocked until we are ready release a new minor. @wulinsheng123 Did you create a PR in the docs repo with the new flag?
ok

File: package.json
Comment (wulinsheng123) [reply package.json:44]: @ematipico do I still revert it?

Comment (ematipico) [reply package.json:44]: The change *makes sense* and I agree with that, but **it should be inside a different PR**. Just that.

Comment (wulinsheng123) [reply package.json:44]: Thank you for resolving the problem.
I will do what you said in the next pr if I encounter the same situation.

**Comment by sarah11918** [comment]
Just checking in on whether this is planned to be accepted.

An issue has been filed in the docs repo simply stating that we will need documentation to accompany this, but no documentation has been proposed. (Issue: https://github.com/withastro/docs/issues/3024 )

 Docs can step in and propose documentation, but we'd rather know that this is going to be accepted first before we do that. :smile:

**Comment by ematipico** [comment]
Happy to merge! @wulinsheng123 could you please update the PR?

**Review by ematipico — APPROVED** [review]

**Comment by matthewp** [comment]
Let's hold off on this one for now. I would like to understand the use-case better. It appears to set some internal code that was never intended to be exposed externally.

**Comment by wulinsheng123** [comment]
> Let's hold off on this one for now. I would like to understand the use-case better. It appears to set some internal code that was never intended to be exposed externally.

Ok, if there is any progress, please tell me because  I need to resolve a conflict.

**Comment by Princesseuh** [comment]
@matthewp Have you had time to revisit this?

**Comment by ematipico** [comment]
> I was probably the one who pushed for it, as I've seen this work on & off in the past in Astro. I think it makes sense to support custom `--mode` so users can e.g. build a site for staging mode.
> 
> I notice in Astro internals, we only differentiate between `development` and `production` mode, so we'd probably have to be careful that e.g. staging mode still builds in `production` mode. At retrospect maybe this should be discussed first 😬

I see the use case, but Astro is a bit different and we need to clearly set user expectations of what a `--mode` is. As far as I understand, `development` mode is the actual dev server, which differs from the actual `production` where the pages are actually **built** into static files (or SSR). 

So if the user set a `--mode=STAGING`, what is it going to affect? What's the expected result in the dev server and in the built artefacts?

**Comment by bluwy** [comment]
Largely (and maybe only?) it'll affect the value of `import.meta.env.MODE` which can be used to render things differently. If `--mode=staging`, I'd expect the dev server to still run in development but with a different `import.meta.env.MODE`, similarly in build it runs in production with a different `import.meta.env.MODE`.

The `mode` name is a bit overloaded with this now though, maybe we should refer `--mode` as `userMode` that's used to achieve `import.meta.env.MODE` and passing that value to Vite's `mode`.

**Comment by matthewp** [comment]
Given there's still some unknowns of how/if this should work, I think this warrants a (probably small) RFC where we work through the ideas. We can then reopen this PR if it turns out to be the approach chosen.

**Pull request closed without merging**
